azkaban-common/src/main/java/azkaban/user/UserUtils.java

```diff
@@ -22,4 +33,56 @@ public static boolean hasPermissionforAction(final UserManager userManager, fina
 
     return false;
   }
+
+  /**
+   * Creates a watch thread which listens to specified files' modification and reloads
+   * configurations
+   */
+  public static void setupWatch(final String fileName, Logger log, ParseConfigFile parser) {
+    Runnable runnable = () -> {
+      WatchService watchService;
+      Path path;
+      try {
+        watchService = FileSystems.getDefault().newWatchService();
+        final File file = new File(fileName);
+        final String dirPath = file.getParent();
+        path = Paths.get(dirPath);
+        path.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+            StandardWatchEventKinds.ENTRY_CREATE);
+      } catch (IOException e) {
+        // Ignore the IOException
+        log.warn("IOException while setting up watch on conf"
+            + e.getMessage());
+        return;
+      }
+      for (;;) {
+        // Watch for modifications
+        WatchKey watchKey;
+        try {
+          watchKey = watchService.take();
+        } catch (InterruptedException ie) {
+          log.warn(ie.getMessage());
+          return;
+        }
+
+        for (WatchEvent<?> event : watchKey.pollEvents()) {
+          // Make sure the modification happened to user xml
+          @SuppressWarnings("unchecked")
+          final Path name = ((WatchEvent<Path>)event).context();
+          final Path child = path.resolve(name);
+          if (!child.toString().equals(fileName)) {
+            continue; // not user xml
```

Comment:
When will this happen?

Reply:
The watch works on the whole directory, however, it is intended for a file in this case. The check skips other files. Moreover, the watch event is triggered even a .swp file is created which needs to be ignored as well.

Reply:
In our production environment, do we expect other files in this directory?